<a href="https://colab.research.google.com/github/cris141187/Notebook/blob/main/Cristiano_8puzzle.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Equipe 
####CRISTIANO DOS SANTOS CARVALHO RA:2920481911028
####GUILHERME SILVA DE OLIVEIRA RA:2920482021009
####ELIANE LEITE ANDRADE RA:2920482021032
####JESSICA DIAS SANTOS CORREIA RA:2920481821042
####SAMUEL SANTANA DIAS DA SILVA RA:2920481821012


In [ ]:
!pip install simpleai

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


# Quebra-cabeça de 8 peças
## Disciplina: Inteligência Artificial e Aplicações - Fatec Ferraz de Vasconcelos
### Professora: Ana Rosa C. Tonão
#### Data: 13/06/2022

In [ ]:
from simpleai.search import astar, SearchProblem

In [ ]:
# Classe contendo métodos para resolver o quebra-cabeça
class PuzzleSolver(SearchProblem):
    # Método de ação para obter a lista dos possíveis
    # números que podem ser movidos para o espaço vazio
    def actions(self, cur_state):
        rows = string_to_list(cur_state)
        row_empty, col_empty = get_location(rows, 'e')

        actions = []
        if row_empty > 0:
            actions.append(rows[row_empty - 1][col_empty])
        if row_empty < 2:
            actions.append(rows[row_empty + 1][col_empty])
        if col_empty > 0:
            actions.append(rows[row_empty][col_empty - 1])
        if col_empty < 2:
            actions.append(rows[row_empty][col_empty + 1])

        return actions

   # Retorna o estado resultante após mover uma peça para o espaço vazio
    def result(self, state, action):
        rows = string_to_list(state)
        row_empty, col_empty = get_location(rows, 'e')
        row_new, col_new = get_location(rows, action)

        rows[row_empty][col_empty], rows[row_new][col_new] = \
                rows[row_new][col_new], rows[row_empty][col_empty]

        return list_to_string(rows)

    # Retorna verdadeiro se um estado for o estado objetivo
    def is_goal(self, state):
        return state == GOAL

    # Retorna uma estimativa da distância de um estado para
    # a meta usando a distância manhattan
    def heuristic(self, state):
        rows = string_to_list(state)

        distance = 0

        for number in '12345678e':
            row_new, col_new = get_location(rows, number)
            row_new_goal, col_new_goal = goal_positions[number]

            distance += abs(row_new - row_new_goal) + abs(col_new - col_new_goal)

        return distance

In [ ]:
# Convert list to string
def list_to_string(input_list):
    return '\n'.join(['-'.join(x) for x in input_list])

# Convert string to list
def string_to_list(input_string):
    return [x.split('-') for x in input_string.split('\n')]

# Encontre a localização 2D do elemento de entrada
def get_location(rows, input_element):
    for i, row in enumerate(rows):
        for j, item in enumerate(row):
            if item == input_element:
                return i, j 

In [ ]:
# Resultado final que queremos alcançar
GOAL = '''1-2-3
4-5-6
7-8-e'''

# Ponto de partida
INITIAL = '''1-e-2
6-3-4
7-5-8'''
INITIAL1 = '''1-2-3
4-5-7
8-6-e'''
INITIAL2 = '''3-4-1
2-5-6
7-8-e'''
INITIAL3 = '''1-3-4
e-5-6
7-8-2'''

In [ ]:
# Crie um cache para a posição de gol de cada peça
goal_positions = {}
rows_goal = string_to_list(GOAL)
for number in '12345678e':
    goal_positions[number] = get_location(rows_goal, number)


### Create the solver object


In [ ]:
result = astar(PuzzleSolver(INITIAL))

In [ ]:
# Imprima os resultados
for i, (action, state) in enumerate(result.path()):
    print()
    if action == None:
        print('Configuração inicial')
    elif i == len(result.path()) - 1:
        print('Depois de mover', action, 'no espaço vazio. Objetivo alcançado!', 'Jogada:', i)
    else:
        print('Depois de mover', action, 'no espaço vazio')

    print(state)



Configuração inicial
1-e-2
6-3-4
7-5-8

Depois de mover 2 no espaço vazio
1-2-e
6-3-4
7-5-8

Depois de mover 4 no espaço vazio
1-2-4
6-3-e
7-5-8

Depois de mover 3 no espaço vazio
1-2-4
6-e-3
7-5-8

Depois de mover 6 no espaço vazio
1-2-4
e-6-3
7-5-8

Depois de mover 1 no espaço vazio
e-2-4
1-6-3
7-5-8

Depois de mover 2 no espaço vazio
2-e-4
1-6-3
7-5-8

Depois de mover 4 no espaço vazio
2-4-e
1-6-3
7-5-8

Depois de mover 3 no espaço vazio
2-4-3
1-6-e
7-5-8

Depois de mover 6 no espaço vazio
2-4-3
1-e-6
7-5-8

Depois de mover 4 no espaço vazio
2-e-3
1-4-6
7-5-8

Depois de mover 2 no espaço vazio
e-2-3
1-4-6
7-5-8

Depois de mover 1 no espaço vazio
1-2-3
e-4-6
7-5-8

Depois de mover 4 no espaço vazio
1-2-3
4-e-6
7-5-8

Depois de mover 5 no espaço vazio
1-2-3
4-5-6
7-e-8

Depois de mover 8 no espaço vazio. Objetivo alcançado! Jogada: 15
1-2-3
4-5-6
7-8-e


In [ ]:
result1 = astar(PuzzleSolver(INITIAL1))

In [ ]:
# Imprima os resultados 1
for i, (action, state) in enumerate(result1.path()):
    print()
    if action == None:
        print('Configuração inicial 1')
    elif i == len(result1.path()) - 1:
        print('Depois de mover', action, 'no espaço vazio. Objetivo alcançado!', 'Jogada:', i)
    else:
        print('Depois de mover', action, 'no espaço vazio')

    print(state)


Configuração inicial 1
1-2-3
4-5-7
8-6-e

Depois de mover 6 no espaço vazio
1-2-3
4-5-7
8-e-6

Depois de mover 8 no espaço vazio
1-2-3
4-5-7
e-8-6

Depois de mover 4 no espaço vazio
1-2-3
e-5-7
4-8-6

Depois de mover 5 no espaço vazio
1-2-3
5-e-7
4-8-6

Depois de mover 7 no espaço vazio
1-2-3
5-7-e
4-8-6

Depois de mover 6 no espaço vazio
1-2-3
5-7-6
4-8-e

Depois de mover 8 no espaço vazio
1-2-3
5-7-6
4-e-8

Depois de mover 7 no espaço vazio
1-2-3
5-e-6
4-7-8

Depois de mover 5 no espaço vazio
1-2-3
e-5-6
4-7-8

Depois de mover 4 no espaço vazio
1-2-3
4-5-6
e-7-8

Depois de mover 7 no espaço vazio
1-2-3
4-5-6
7-e-8

Depois de mover 8 no espaço vazio. Objetivo alcançado! Jogada: 12
1-2-3
4-5-6
7-8-e


In [ ]:
result2 = astar(PuzzleSolver(INITIAL2))

In [ ]:
# Imprima os resultados 2
for i, (action, state) in enumerate(result2.path()):
    print()
    if action == None:
        print('Configuração inicial 2')
    elif i == len(result2.path()) - 1:
        print('Depois de mover', action, 'no espaço vazio. Objetivo alcançado!', 'Jogada:', i)
    else:
        print('Depois de mover', action, 'no espaço vazio')

    print(state)


Configuração inicial 2
3-4-1
2-5-6
7-8-e

Depois de mover 8 no espaço vazio
3-4-1
2-5-6
7-e-8

Depois de mover 5 no espaço vazio
3-4-1
2-e-6
7-5-8

Depois de mover 2 no espaço vazio
3-4-1
e-2-6
7-5-8

Depois de mover 3 no espaço vazio
e-4-1
3-2-6
7-5-8

Depois de mover 4 no espaço vazio
4-e-1
3-2-6
7-5-8

Depois de mover 1 no espaço vazio
4-1-e
3-2-6
7-5-8

Depois de mover 6 no espaço vazio
4-1-6
3-2-e
7-5-8

Depois de mover 2 no espaço vazio
4-1-6
3-e-2
7-5-8

Depois de mover 3 no espaço vazio
4-1-6
e-3-2
7-5-8

Depois de mover 4 no espaço vazio
e-1-6
4-3-2
7-5-8

Depois de mover 1 no espaço vazio
1-e-6
4-3-2
7-5-8

Depois de mover 3 no espaço vazio
1-3-6
4-e-2
7-5-8

Depois de mover 2 no espaço vazio
1-3-6
4-2-e
7-5-8

Depois de mover 6 no espaço vazio
1-3-e
4-2-6
7-5-8

Depois de mover 3 no espaço vazio
1-e-3
4-2-6
7-5-8

Depois de mover 2 no espaço vazio
1-2-3
4-e-6
7-5-8

Depois de mover 5 no espaço vazio
1-2-3
4-5-6
7-e-8

Depois de mover 8 no espaço vazio. Objetivo alcançado! J

In [ ]:
result3 = astar(PuzzleSolver(INITIAL3))

In [ ]:
# Imprima os resultados 3
for i, (action, state) in enumerate(result3.path()):
    print()
    if action == None:
        print('Configuração inicial 3')
    elif i == len(result3.path()) - 1:
        print('Depois de mover', action, 'no espaço vazio. Objetivo alcançado!', 'Jogada:', i)
    else:
        print('Depois de mover', action, 'no espaço vazio')

    print(state)


Configuração inicial 3
1-3-4
e-5-6
7-8-2

Depois de mover 1 no espaço vazio
e-3-4
1-5-6
7-8-2

Depois de mover 3 no espaço vazio
3-e-4
1-5-6
7-8-2

Depois de mover 4 no espaço vazio
3-4-e
1-5-6
7-8-2

Depois de mover 6 no espaço vazio
3-4-6
1-5-e
7-8-2

Depois de mover 2 no espaço vazio
3-4-6
1-5-2
7-8-e

Depois de mover 8 no espaço vazio
3-4-6
1-5-2
7-e-8

Depois de mover 5 no espaço vazio
3-4-6
1-e-2
7-5-8

Depois de mover 4 no espaço vazio
3-e-6
1-4-2
7-5-8

Depois de mover 3 no espaço vazio
e-3-6
1-4-2
7-5-8

Depois de mover 1 no espaço vazio
1-3-6
e-4-2
7-5-8

Depois de mover 4 no espaço vazio
1-3-6
4-e-2
7-5-8

Depois de mover 2 no espaço vazio
1-3-6
4-2-e
7-5-8

Depois de mover 6 no espaço vazio
1-3-e
4-2-6
7-5-8

Depois de mover 3 no espaço vazio
1-e-3
4-2-6
7-5-8

Depois de mover 2 no espaço vazio
1-2-3
4-e-6
7-5-8

Depois de mover 5 no espaço vazio
1-2-3
4-5-6
7-e-8

Depois de mover 8 no espaço vazio. Objetivo alcançado! Jogada: 17
1-2-3
4-5-6
7-8-e


### Computation time of heuristic funcion

In [ ]:
%%timeit
astar(PuzzleSolver(INITIAL))


1.24 s ± 80.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [ ]:
%%timeit
astar(PuzzleSolver(INITIAL1))

126 ms ± 2.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [ ]:
%%timeit
astar(PuzzleSolver(INITIAL2))

4.14 s ± 224 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [ ]:
%%timeit
astar(PuzzleSolver(INITIAL3))

4.29 s ± 75.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


# Respostas:

**5. Compare também os tempos de execução.**

##Resposta:
###1-Estado Inicial Padrão (1-e-2-6-3-4-7-5-8)
#####Temos o tempo de execução em:
#####1.24 s ± 80.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
###2-Estado Inicial 1 (1-2-3-4-5-7-8-6-e)
#####Temos o tempo de execução em:
#####126 ms ± 1.27 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

###3-Estado Inicial 2 (3-4-1-2-5-6-7-8-e)
#####Temos o tempo de execução em:
#####4.06 s ± 38.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

###4-Estado Inicial 3 (1-3-4-e-5-6-7-8-2)
#####Temos o tempo de execução em:
#####4.29 s ± 75.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)







**6. Compare a quantidade de iterações (jogadas) necessária para atingir o objetivo em cada estado inicial.**

##Resposta: 
####Objetivo alcançado no estado inicial (1-e-2-6-3-4-7-5-8) com Jogadas = 15
####Objetivo alcançado no estado inicial (1-2-3-4-5-7-8-6-e) com Jogadas = 12
####Objetivo alcançado no estado inicial (3-4-1-2-5-6-7-8-e) com Jogadas = 18
####Objetivo alcançado no estado inicial (1-3-4-e-5-6-7-8-2) com Jogadas = 17

## FIM